echopop/nwfsc_feat/stratified.py

```diff
@@ -0,0 +1,1002 @@
+        used in acoustic survey analysis, where variance is weighted by
+        the square of the distance weights.
+        """
+
+        # Convert values to per-distance basis, handling division by zero
+        with np.errstate(divide="ignore", invalid="ignore"):
+            values_adjusted = awk.where(
+                sampled_distances != 0, sampled_values / sampled_distances, 0.0
+            )
+
+        # Calculate squared deviations from mean
+        squared_deviation = (values_adjusted - mean_values[..., None]) ** 2
+
+        # Apply distance weights squared
+        weighted_squared_deviation = awk.sum(stratified_weights**2 * squared_deviation, axis=-1)
+
+        # Return variance (weighted sum divided by DOF)
+        return weighted_squared_deviation / awk.Array(sample_dof)[:, None]
+
+    def stratified_bootstrap(
+        self,
+        data_df: pd.DataFrame,
+        stratify_by: List[str],
+        variable: str,
+    ) -> None:
+        """
+        Perform stratified bootstrap resampling analysis.
+
+        Executes the complete bootstrap analysis pipeline including transect
+        summarization, stratum summarization, bootstrap sample generation,
+        and computation of bootstrap statistics.
+
+        Parameters
+        ----------
+        data_df : pd.DataFrame
+            Input DataFrame containing virtual transect data with all required
+            columns for analysis.
+        stratify_by : List[str]
+            List of column names defining the stratification (e.g., ['geostratum_inpfc']).
+        variable : str
+            Name of the response variable column (e.g., 'biomass', 'abundance').
+
+        Notes
+        -----
+        This method populates the resampled_df attribute with bootstrap replicates
+        organized by stratum and replicate. The bootstrap sampling is done without
+        replacement within each stratum.
+
+        The method performs the following steps:
+        1. Summarize transect-level data
+        2. Summarize stratum-level data
+        3. Generate bootstrap sample arrays
+        4. Compute distance-weighted means and variances
+        5. Format results into structured DataFrame
+        """
+
+        # Store the response variable name
+        self.variable = variable
+
+        # Summarize transect data within strata
+        self._summarize_transects(data_df, stratify_by)
+
+        # Summarize at stratum level
+        self._summarize_strata(stratify_by)
+
+        # Summarize at the survey level
+        self._summarize_survey()
+
+        # Calculate degrees of freedom for variance calculations
+        sample_dof = self._dof()
+
+        # Generate bootstrap sample arrays
+        sampled_distances, sampled_areas, sampled_values = self._prepare_bootstrap_arrays()
+
+        # Compute distance-based stratified weights, handling division by zero
+        mean_distances = awk.mean(sampled_distances, axis=-1, keepdims=True)
```

Comment:
Is the use of awkward arrays here (and in other places) to avoid having to do list comprehension (loop)? Am trying to understand why, since the calculated results are then converted back to numpy and eventually stored in a dataframe.